File: test_17_18/caching.py
import threading
import time
import random
import requests
from flask import Flask, request, jsonify

class SimpleCache:
    def __init__(self):
        self.data = {}

    def get(self, key):
        return self.data.get(key)

    def set(self, key, val):
        self.data[key] = val

app = Flask(__name__)
cache = SimpleCache()

failed_logins = {}

@app.route('/login', methods=['POST'])
def login():
    username = request.form['username']
    ip = request.remote_addr
    now = int(time.time())

    if username not in failed_logins:
        failed_logins[username] = []

    failed_logins[username].append(now)

    if len(failed_logins[username]) > 6:
        print(f"Potential abuse: {username} at {ip}")

    return jsonify({"error": "Invalid credential"}), 401

@app.route('/reset-password', methods=['POST'])
def reset_password():
    username = request.form['username']
    ip = request.remote_addr

    return jsonify({"status": f"Reset link sent for {username}"}), 200

def withdrawal_api(user, amount):
    time.sleep(random.uniform(0.1, 0.2))
    return {"result": "success", "msg": f"Withdrew {amount}"}

def fetch_exchange_rate():
    url = "https://dummy-api.com/exchange"
    try:
        resp = requests.get(url)
        return resp.json()
    except Exception:
        return {"rate": 1.23}

def action_thread(user):
    for _ in range(3):
        fetch_exchange_rate()
        withdrawal_api(user, random.randint(100, 1000))
        time.sleep(0.06)

def main():
    thds = []
    for i in range(7):
        t = threading.Thread(target=action_thread, args=(f"user{i}",))
        thds.append(t)
        t.start()
    for t in thds:
        t.join()

if __name__ == "__main__":
    main()

File: test_17_18/caching.java
import java.util.*;
import java.io.*;
import java.net.*;
import java.util.concurrent.*;

class SimpleCache {
    private final Map<String, Object> data = new HashMap<>();
    public Object get(String key) { return data.get(key); }
    public void set(String key, Object val) { data.put(key, val); }
}

class AuthHandler {
    private static final Map<String, List<Long>> loginAttempts = new HashMap<>();

    public static String login(String username, String ip) {
        long now = System.currentTimeMillis();
        if (!loginAttempts.containsKey(username)) {
            loginAttempts.put(username, new ArrayList<>());
        }
        loginAttempts.get(username).add(now);
        if (loginAttempts.get(username).size() > 6) {
            System.out.println("Potential brute-force: " + username + " from " + ip);
        }
        return "{\"error\": \"Invalid credential\"}";
    }

    public static String resetPassword(String username, String ip) {
        return "{\"status\": \"Reset link sent for " + username + "\"}";
    }
}

class WithdrawalService {
    public static String withdraw(String user, int amount) {
        try { Thread.sleep((int)(100 + Math.random()*70)); } catch(InterruptedException e){}
        return "Withdrew " + amount;
    }
}

class ExchangeRateClient {
    public static double fetchRate() {
        try {
            URL url = new URL("https://dummy-api.com/exchange");
            BufferedReader in = new BufferedReader(new InputStreamReader(url.openStream()));
            String inputLine;
            StringBuilder sb = new StringBuilder();
            while ((inputLine = in.readLine()) != null)
                sb.append(inputLine);
            in.close();
            return 1.23;
        } catch (Exception e) {
            return 1.23;
        }
    }
}

class CircuitBreakerExample implements Runnable {
    private final String user;
    public CircuitBreakerExample(String user) { this.user = user; }
    public void run() {
        for (int i = 0; i < 3; i++) {
            ExchangeRateClient.fetchRate();
            WithdrawalService.withdraw(user, (int)(100 + Math.random()*1000));
            try { Thread.sleep(50); } catch (InterruptedException e) {}
        }
    }
}

public class MainApp {
    public static void main(String[] args) throws Exception {
        ExecutorService exec = Executors.newFixedThreadPool(7);
        for (int i = 0; i < 7; i++) {
            exec.submit(new CircuitBreakerExample("user" + i));
        }
        exec.shutdown();
        exec.awaitTermination(2, TimeUnit.MINUTES);
    }
}
